### PR TITLE
upgrade Bootstrap to 4

### PR DESCRIPTION
**Pull request by keshav234156**
#508 
Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [ ] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `grunt jasmine`
* [ ] code is in uniquely-named feature branch and has no merge conflicts
* [ ] PR is descriptively titled
* [ ] PR body includes `fixes #0000`-style reference to original issue #
* [ ] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software

We have a loose schedule of reviewing and pulling in changes every Tuesday and Friday, and publishing changes on Fridays. 

Thanks!

### CODE REVIEW

**Comment by keshav234156** [comment]
@sagarpreet-chadha @NitinBhasneria @shreyaa-s @Shulammite-Aso Please review. We can leave the passing of jest-puppeteer test as of now. The test are correct and are passing locally but it look like travis run with --runInBand to fasten te process which fails the test.locally it can be simulated by ```node ./node_modules/.bin/jest --runInBand``` . I  have also added changes to Gruntfile.js(made it as it was prior to #519 ) and added it to eslintignore as process were not running related to grunt because of it.

**Comment by keshav234156** [comment]
Also, I have made this new PR (previous PR #561) because my previous forked repo was having too much problem with test and all branches were messed up (I don't why :confused: ?).

**Comment by sagarpreet-chadha** [comment]
@keshav234156 , did you find any way we can remove this param from travis so that the tests can pass? Thanks!

**Comment by keshav234156** [comment]
@sagarpreet-chadha Not yet!!

**Comment by shreyaa-sharmaa** [comment]
I can replicate it locally but why is it triggered on selected PRs? And persists even on restarting travis? Do you have any idea why is it so @keshav234156?

**Comment by jywarren** [comment]
If you get really stuck here, perhaps you might reach out to `icarito` or `alaxalves`! Thanks for working hard on this, and if at all possible I recommend following conventions established by Travis and Jest, and linking to them so folks can follow along. Thank you!

**Comment by keshav234156** [comment]
@shreyaa-s Can you please look at these test as well !!

**Comment by shreyaa-sharmaa** [comment]
I tried but for this one the last line, i.e. ` await page.click('.woofmark-mode-wysiwyg');` isn't working. The mode doesn't change. The safeguard added for that in the starting of test works. But in this case the starting of the text gets selected and is changed to bold/italic that's why it couldn't find the string specified since we format the existing selected text, the test only fails for whichever text comes second. There are a few options we could try:
1. Converting to md and checking for `*` or `_`.
2. Adding `await page.evaluate(() => document.querySelector('.wk-wysiwyg').textContent += ' ');` but this makes the rest of the text loose all formatting, like this:
italic
3. backspace the selected text.
I wasn't able to figure out why the change mode button didn't work, if anyone can that would be great help. And if we can implement anyone one of these, which one should it be?

**Comment by jywarren** [comment]
Hi @keshav234156 just wondering if you had a chance to try one of @shreyaa-sharmaa's ideas above? Did any of them work?

**Comment by keshav234156** [comment]
fixed done for bold and italic by backspace. Now need to find a fix for center text.

**Comment by keshav234156** [comment]
fixed center test as well!!
@jywarren Please review.

File: test/ui-testing/bold.test.js [hunk 1]
```diff
@@ -7,11 +7,23 @@ beforeAll(async () => {
 
 describe('Bold Text', () => {
   test('Adds strong text in rich text mode', async () => {
+    // switches to wysiwyg mode if it is in markdown mode
+    await page.screenshot({path: 'buddy-screenshot1.png'});
```
[the remark this answers is not in the document]
Comment (sagarpreet-chadha) [reply test/ui-testing/bold.test.js:11]: Correct!

**Comment by jywarren** [comment]
Hi, can you resolve conflicts here? I'd love to include this in `v2.2.0`!

**Comment by jywarren** [comment]
Great hard work on the testing issues, as well! 🎉

**Comment by keshav234156** [comment]
@jywarren Ok doing it now!

**Comment by keshav234156** [comment]
@jywarren Can you please see why test are not starting

**Comment by jywarren** [comment]
Hi, @keshav234156 - one thing that can happen sometimes is that it's easier to just rebase the whole thing cleanly to get Travis to restart. It looks like there's something odd going on here even with GitPod leaving so many comments. Can you try a clean rebase?

**Review by Shulammite-Aso — APPROVED** [review]

File: Gruntfile.js [hunk 2]
```diff
@@ -1,81 +1,84 @@
 module.exports = function(grunt) {
-  grunt.loadNpmTasks('grunt-browserify');
 
-  require('matchdep').filterDev('grunt-*').forEach(grunt.loadNpmTasks);
+    grunt.loadNpmTasks('grunt-browserify');
```
Comment (jywarren) [inline Gruntfile.js:3]: Uh oh - were these commits supposed to make it into here? Just checking!

Comment (jywarren) [reply Gruntfile.js:3]: Ah, i see here! all good then! https://github.com/publiclab/PublicLab.Editor/pull/566#issuecomment-660536609

**Comment by jywarren** [comment]
Hi! Almost there -- i see a few small issues!

I see a small issue with the "Help" icon here:

image

And the `Use current location` button needs an outline, as do a few other buttons. I think they work better with outlines otherwise they are hard to see as buttons, you know?

image

image

The history button too:

image

@keshav234156 I think this will be all good once you can make these tweaks, thank you so much!!!

**Comment by jywarren** [comment]
Hi @keshav234156 i hope you're well! Do you think you'll have any time to wrap this issue up, or do you think it'd be best if someone else took a look at it? Thank you!!!

**Comment by Sagarpreet** [comment]
Moved here: https://github.com/publiclab/PublicLab.Editor/pull/636
Thanks @keshav234156 💯

**Pull request closed without merging**